sharding/shard.go

```diff
@@ -0,0 +1,262 @@
+	} else {
+		encoded = []byte{0}
+	}
+	if err := s.shardDB.Put(key, encoded); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SaveHeader adds the collation header to shardDB.
+func (s *Shard) SaveHeader(header *CollationHeader) error {
+	// checks if the header has a chunk root set before saving.
+	if header.ChunkRoot() == nil {
+		return fmt.Errorf("header needs to have a chunk root set before saving")
+	}
+
+	encoded, err := header.EncodeRLP()
+	if err != nil {
+		return fmt.Errorf("cannot encode header: %v", err)
+	}
+
+	// uses the hash of the header as the key.
+	if err := s.shardDB.Put(header.Hash(), encoded); err != nil {
+		return fmt.Errorf("cannot update shardDB: %v", err)
+	}
+	return nil
+}
+
+// SaveBody adds the collation body to the shardDB and sets availability.
+func (s *Shard) SaveBody(body []byte) error {
+	// TODO: check if body is empty and throw error.
+	// TODO: dependent on blob serialization.
+	// chunkRoot := getChunkRoot(body) using the blob algorithm utils.
+	// right now we will just take the raw keccak256 of the body until #92 is merged.
+	chunkRoot := common.BytesToHash(body)
+
+	if err := s.shardDB.Put(chunkRoot, body); err != nil {
+		return err
+	}
+	s.SetAvailability(&chunkRoot, true)
+	return nil
+}
+
+// SaveCollation adds the collation's header and body to shardDB.
+func (s *Shard) SaveCollation(collation *Collation) error {
+	if err := s.ValidateShardID(collation.Header()); err != nil {
+		return err
+	}
+	if err := s.SaveHeader(collation.Header()); err != nil {
+		return err
+	}
+	if err := s.SaveBody(collation.Body()); err != nil {
+		return err
+	}
+	return nil
+}
+
+// SetCanonical sets the collation header as canonical in the shardDB. This is called
+// after the period is over and over 2/3 notaries voted on the header.
+func (s *Shard) SetCanonical(header *CollationHeader) error {
+	if err := s.ValidateShardID(header); err != nil {
+		return err
+	}
+	// the header needs to have been stored in the DB previously, so we
+	// fetch it from the shardDB.
+	hash := header.Hash()
+	dbHeader, err := s.HeaderByHash(&hash)
+	if err != nil {
+		return err
+	}
+
+	// checks if the header has a corresponding body in the DB.
+	body, err := s.BodyByChunkRoot(dbHeader.ChunkRoot())
+	if err != nil {
+		return fmt.Errorf("no corresponding collation body saved in shardDB: %v", body)
+	}
+
+	key := canonicalCollationLookupKey(dbHeader.ShardID(), dbHeader.Period())
+	encoded, err := dbHeader.EncodeRLP()
+	if err != nil {
+		return fmt.Errorf("cannot encode header: %v", err)
+	}
+	// sets the key to be the canonical collation lookup key and val as RLP encoded
+	// collation header.
+	if err := s.shardDB.Put(key, encoded); err != nil {
+		return err
+	}
+	return nil
+}
+
+// dataAvailabilityLookupKey formats a string that will become a lookup
+// key in the shardDB.
+func dataAvailabilityLookupKey(chunkRoot *common.Hash) common.Hash {
+	key := fmt.Sprintf("availability-lookup:%s", chunkRoot.String())
```

Comment:
I don't think you need to call `.String()` it should be automatic for the formatter with `%s`.